src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -1991,6 +1991,42 @@ internal PointerTypeSymbol CreatePointerTypeSymbol(TypeSymbol elementType, Nulla
             return new PointerTypeSymbol(TypeWithAnnotations.Create(elementType, elementNullableAnnotation));
         }
 
+        internal FunctionPointerTypeSymbol CreateFunctionPointerTypeSymbol(
+            TypeWithAnnotations returnType,
+            ImmutableArray<TypeWithAnnotations> parameterTypes,
+            RefKind returnRefKind,
+            ImmutableArray<RefKind> parameterRefKinds)
+        {
+            if (returnType.IsDefault)
+            {
+                throw new ArgumentNullException(nameof(returnType));
+            }
+
+            if (parameterTypes.IsDefault || parameterTypes.Any(param => param.IsDefault))
+            {
+                throw new ArgumentNullException(nameof(parameterTypes));
+            }
+
+            if (parameterRefKinds.IsDefault)
+            {
+                throw new ArgumentNullException(nameof(parameterRefKinds));
+            }
+
+            if (parameterRefKinds.Length != parameterTypes.Length)
+            {
+                // Given {0} parameter types and {1} parameter ref kinds. These must be the same.
+                throw new ArgumentException(string.Format(CSharpResources.NotSameNumberParameterTypesAndRefKinds, parameterTypes.Length, parameterRefKinds.Length));
+            }
+
+            if (returnRefKind == RefKind.Out)
+            {
+                //'RefKind.Out' is not a valid ref kind for a return type.
+                throw new ArgumentException(CSharpResources.OutIsNotValidForReturn);
```

Comment:
If somebody writes `delegate*<out int>` today, is the existing symbol going to still mark that as out or is it not even going to make it into the symbol model? My concern being this is fine as long as it's blocked from ever appearing in any symbol anywhere, not just through callers of this API.

Reply:
I'm not sure what the public api would do, actually, it's a good test to add.